File: Java/Leetcode/src/Medium/LexicographicalNumbers.java
package Medium;

import java.util.ArrayList;
import java.util.List;

public class LexicographicalNumbers {

	List<Integer> result;
	int n;
	
	public List<Integer> lexicalOrder(int n) {
        this.result = new ArrayList<Integer>();
        this.n = n;
        
        for(int i = 1; i <= 9; i++) {
        	dfs(i);
        }
        
        return this.result;
    }
	
	public void dfs(int currentNum) {
		if(currentNum > this.n) {
			return;
		}
		this.result.add(currentNum);
		
		
		for(int i = 0; i <= 9; i++) {
			int nextNum = currentNum * 10 + i;
			if(nextNum > n) {
				break;
			}
			
			dfs(nextNum);
			
		}
	}
	
	public static void main(String[] args) {
		LexicographicalNumbers lexicographicalNumbers = new LexicographicalNumbers();
		
		System.out.println(lexicographicalNumbers.lexicalOrder(13));
		System.out.println(lexicographicalNumbers.lexicalOrder(2));

	}

}

File: Python/Leetcode/Medium/Lexicographical Numbers.py
class Solution:
    def lexicalOrder(self, n: int) -> list[int]:
        result = []
        def dfs(currentNum):
            # n보다 클경우 현제 숫자 종료
            if currentNum > n:
                return
            result.append(currentNum) # n이하일 경우 추가 한다.
            for i in range(10):
                nextNum = currentNum * 10 + i # 다음수는 10을 곱하고 i를 더한다.
                # 다음 수가 N보다 클경우 종료
                if nextNum > n:
                    break
                # n이하일 경우 다음 dfs 탐색
                dfs(nextNum)

        # 1부터 9까지 순서대로 탐색한다.
        for i in range(1, 10):
            dfs(i)

        return result

solution = Solution()
print(solution.lexicalOrder(13))
print(solution.lexicalOrder(2))